codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -284,194 +316,542 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
         }
 
         @Override
-        public boolean isEndStream() {
-            return frame.isEndStream();
+        protected SocketAddress remoteAddress0() {
+            return parent().remoteAddress();
         }
 
         @Override
-        public int padding() {
-            return frame.padding();
+        protected void doBind(SocketAddress localAddress) throws Exception {
+            throw new UnsupportedOperationException();
         }
 
         @Override
-        public Http2StreamFrame streamId(int streamId) {
-            return frame.streamId(streamId);
+        protected void doDisconnect() throws Exception {
+            doClose();
         }
 
         @Override
-        public int streamId() {
-            return frame.streamId();
+        protected void doClose() throws Exception {
+            closed = true;
+
+            if (!streamClosedWithoutError && isStreamIdValid(stream().id())) {
+                Http2StreamFrame resetFrame = new DefaultHttp2ResetFrame(Http2Error.CANCEL).stream(stream());
+                ctx.writeAndFlush(resetFrame);
+            }
+
+            while (!inboundBuffer.isEmpty()) {
+                ReferenceCountUtil.release(inboundBuffer.poll());
+            }
         }
 
         @Override
-        public String name() {
-            return frame.name();
+        protected void doBeginRead() {
+            if (readInProgress) {
+                return;
+            }
+
+            final RecvByteBufAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
+            allocHandle.reset(config());
+            if (inboundBuffer.isEmpty()) {
+                readInProgress = true;
+                return;
+            }
+
+            do {
+                Object m = inboundBuffer.poll();
+                if (m == null) {
+                    break;
+                }
+                if (!doRead0(m, allocHandle)) {
+                    // Channel closed, and already cleaned up.
+                    return;
+                }
+            } while (allocHandle.continueReading());
+
+            allocHandle.readComplete();
+            pipeline().fireChannelReadComplete();
+        }
+
+        @Override
+        protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+            if (closed) {
+                throw CLOSED_CHANNEL_EXCEPTION;
+            }
+
+            try {
+                if (sizeEstimatorHandle == null) {
+                    sizeEstimatorHandle = config().getMessageSizeEstimator().newHandle();
+                }
+                for (;;) {
+                    final Object msg = in.current();
+                    if (msg == null) {
+                        break;
+                    }
+                    if (msg == REEVALUATE_WRITABILITY_MESSAGE) {
+                        in.remove();
+                        continue;
+                    }
+
+                    final int bytes = sizeEstimatorHandle.size(msg);
+
+                    /*
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)